lib/spack/spack/cmd/env.py

```diff
@@ -22,57 +22,765 @@
+
+
+def repair(environment_name):
+    """Recovers from crash during critical section of write().
+    Possibilities:
+
+        tmp_new, dest
+        tmp_new, tmp_old
+        tmp_old, dest
+    """
+    tmp_new, dest, tmp_old = get_write_paths(get_env_root(environment_name))
+    if os.path.exists(tmp_old):
+        if not os.path.exists(dest):
+            shutil.move(tmp_new, dest)
+        else:
+            shutil.rmtree(tmp_old)
+        tty.info("Previous update completed")
+    elif os.path.exists(tmp_new):
+        tty.info("Previous update did not complete")
+    else:
+        tty.info("Previous update may have completed")
+
+    if os.path.exists(tmp_new):
+        shutil.rmtree(tmp_new)
+
+
+def read(environment_name):
+    # Check that env is in a consistent state on disk
+    env_root = get_env_root(environment_name)
+
+    # Read env.yaml file
+    env_yaml = spack.config._read_config_file(
+        fs.join_path(env_root, 'env.yaml'),
+        spack.schema.env.schema)
+
+    dotenv_dir = get_dotenv_dir(env_root)
+    with open(fs.join_path(dotenv_dir, 'environment.json'), 'r') as F:
+        environment_dict = sjson.load(F)
+    environment = Environment.from_dict(environment_name, environment_dict)
+    if env_yaml:
+        environment.yaml = env_yaml['env']
+
+    return environment
+
+
+# =============== Modifies Environment
+
+def environment_create(args):
+    if os.path.exists(get_env_root(args.environment)):
+        raise tty.die("Environment already exists: " + args.environment)
+
+    _environment_create(args.environment)
+
+
+def _environment_create(name):
+    environment = Environment(name)
+    write(environment)
+
+
+def environment_add(args):
+    check_consistent_env(get_env_root(args.environment))
+    environment = read(args.environment)
+    parsed_specs = spack.cmd.parse_specs(args.package)
+
+    if args.all:
+        # Don't allow command-line specs with --all
+        if len(parsed_specs) > 0:
+            tty.die('Cannot specify --all and specs too on the command line')
+
+        yaml_specs = environment.yaml['specs']
+        if len(yaml_specs) == 0:
+            tty.msg('No specs to add from env.yaml')
+
+        # Add list of specs from env.yaml file
+        for user_spec in yaml_specs:
+            environment.add(user_spec.format(), report_existing=False)
+    else:
+        for spec in parsed_specs:
+            environment.add(spec.format())
+
+    write(environment)
+
+
+def environment_remove(args):
+    check_consistent_env(get_env_root(args.environment))
+    environment = read(args.environment)
+    if args.all:
+        environment.clear()
     else:
-        spec = args.spec[0]
-        cmd = args.spec[1:]
+        for spec in spack.cmd.parse_specs(args.package):
+            environment.remove(spec.format())
+    write(environment)
+
+
+def environment_spec(args):
+    environment = read(args.environment)
+    prepare_repository(environment)
+    prepare_config_scope(environment)
```

Comment:
@citibeth I assumed your intention was to add a command which works like `spack spec` but uses the environment (and in the future optionally the repository) of the environment. I added these commands to produce that effect.